### PR TITLE
[DO NOT MERGE] Functional test runner refactor

### PR DESCRIPTION
**Pull request by stevenvergenz**
A new test runner setup. Please ignore changes to the individual tests, they're not ported yet.

image

New test paradigm:
* Tests run indefinitely, until stopped by the runner. This means that state-cycling tests should reset at the end of their sequence.
* Tests are expected to not exceed the volume of 2x1x2 meters around the origin.
* Tests should avoid spawning labels to document state changes. Instead, use `app.setOverrideText`.

### CODE REVIEW

**Comment by eanders-MS** [comment]
>         const updateLabel = (lightType: string) => {

Light-test updates its message label. Is this something `expectedResultDescription` could support?

---
Refers to: packages/functional-tests/src/tests/light-test.ts:41 in 106c134. [](commit_id = 106c134066caa338e7139ea20b665e95d3d44220, deletion_comment = False)

**Comment by eanders-MS** [comment]
>     private running = true;

Some of the tests I've written have this `running` field that gets checked periodically to see if the test should continue or not. Is there something in the test base class that could be queried instead?

---
Refers to: packages/functional-tests/src/tests/interpolation-test.ts:12 in 106c134. [](commit_id = 106c134066caa338e7139ea20b665e95d3d44220, deletion_comment = False)

File: packages/functional-tests/tsconfig.json [hunk 1]
```diff
@@ -1,8 +1,18 @@
 {
   "extends": "../tsconfig.packages.json",
-  "include": [
-    "./src"
+  "files": [
```
Comment (eanders-MS) [inline packages/functional-tests/tsconfig.json]: Why this change? I want to avoid the use of the "files" section. It's one more thing to maintain. Was "include": [ "./src" ] not working for you?

File: packages/functional-tests/src/tests/failure-test.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Test } from '../test';
+
+export default class FailureTest extends Test {
+    public expectedResultDescription = "Fails";
```
Comment (eanders-MS) [inline packages/functional-tests/src/tests/failure-test.ts]: Name seems overly specific. How about just `statusMessage`?

File: packages/functional-tests/tsconfig.json (hunk 1 above)
Comment (stevenvergenz) [reply packages/functional-tests/tsconfig.json]: This is strictly temporary. I didn't want it building files I hadn't fixed yet.

**Comment by stevenvergenz** [comment]
>         const updateLabel = (lightType: string) => {

Yes, `this.app.setOverrideText()`

---
In reply to: [465689656](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/208#issuecomment-465689656) [](ancestors = 465689656)

---
Refers to: packages/functional-tests/src/tests/light-test.ts:41 in edfeb35. [](commit_id = edfeb359510a839091b65f095487893aed44f3ec, deletion_comment = False)

**Comment by stevenvergenz** [comment]
>     private running = true;

Yes, those can be replaced with `this.stopped` now. I'm still working on updating all the tests.

---
In reply to: [465690010](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/208#issuecomment-465690010) [](ancestors = 465690010)

---
Refers to: packages/functional-tests/src/tests/interpolation-test.ts:12 in edfeb35. [](commit_id = edfeb359510a839091b65f095487893aed44f3ec, deletion_comment = False)

File: packages/functional-tests/src/app.ts [hunk 3]
```diff
@@ -3,203 +3,252 @@
...
-    private activeTests: { [id: string]: Test } = {};
     private _rpc: MRERPC.ContextRPC;
-    private firstUser: MRESDK.User;
-    private _connectedUsers: { [id: string]: MRESDK.User } = {};
+    private firstUser: MRE.User;
+    private _connectedUsers: { [id: string]: MRE.User } = {};
+    private testNames = Object.keys(Factories).sort();
+    private testResults: { [name: string]: boolean } = {};
+
+    private activeTest: Test = null;
```
Comment (eanders-MS) [inline packages/functional-tests/src/app.ts]: >activeTest [](start = 12, length = 10)

As we discussed offline: When the functional-test app is restarted and reloads into the scene, we need to be able to immediately click on the same point in space to relaunch the same test as before the restart. This is important for rapid dev/test iteration when developing a test. Is that still supported?

Comment (stevenvergenz) [reply packages/functional-tests/src/app.ts]: The query argument `test` will skip to that item in the menu, so yes. There are also `autorun` and `nomenu` arguments available.

Comment (eanders-MS) [reply packages/functional-tests/src/app.ts]: That isn't the same. I don't want the test to auto start.

---
In reply to: [258618909](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/208#discussion_r258618909) [](ancestors = 258618909)

Comment (eanders-MS) [reply packages/functional-tests/src/app.ts]: In a single session, we must be able to:
- click on the same spot in space to start the test after restarting the app.
- explicitly start the test (not auto start).
- run the test multiple times.

---
In reply to: [258619934](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/208#discussion_r258619934) [](ancestors = 258619934,258618909)

Comment (stevenvergenz) [reply packages/functional-tests/src/app.ts]: Maybe I need button labels. Set your URI to `ws://localhost:3901/?test=my-test`, and the app will start on your test, ready to go when you click Start (the bottom button).

---
In reply to: [258620721](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/208#discussion_r258620721) [](ancestors = 258620721,258619934,258618909)

**Comment by stevenvergenz** [comment]
Per our face-to-face meeting, I will redesign the menu to allow skipping to an arbitrary test in one or two clicks. The challenge will be keeping it compact. Will experiment.
